### PR TITLE
Optimize the time complexity of MMR diversifier

### LINKED ISSUE #65: [demo] MMR Diverisifier based LinkList
Update the implementation of `MaximalMarginalRelevanceDiversifier`, from the original `List` based implementation to the `LinkList` based implementation.

### PR DESCRIPTION
**Pull request by Liusy12138**
Update `MaximalMarginalRelevanceDiversifier`, use a customized LinkedList data structure instead of Java ArrayList to find best MMR node of the input List.

close #65 .

### CODE REVIEW

File: demo/movielens/online/src/main/java/com/dmetasoul/metaspore/demo/movielens/diversify/diversifier/impl/MaximalMarginalRelevanceDiversifier.java [hunk 1]
```diff
@@ -29,23 +29,27 @@
...
         if (lambda == null) {
             lambda = DEFAULT_LAMBDA;
         }
+        LinkList itemLinkList = new LinkList();
         int genreCount = Utils.groupByType(itemModels).size();
         if (window == null || window > genreCount) {
             window = genreCount;
         }
 
+        itemLinkList.addAll(itemModels);
```
Comment (intelligencegear) [inline demo/movielens/online/src/main/java/com/dmetasoul/metaspore/demo/movielens/diversify/diversifier/impl/MaximalMarginalRelevanceDiversifier.java:51]: please merge this line with the line 45 together.

File: demo/movielens/online/src/main/java/com/dmetasoul/metaspore/demo/movielens/diversify/diversifier/impl/MaximalMarginalRelevanceDiversifier.java [hunk 2]
```diff
@@ -142,6 +150,67 @@ public static Double getSimScore(ItemModel item, HashMap<String, Integer> itemIn
...
+    class LinkList {
+        int size;
+        ListNode head;
+
+        public LinkList() {
+            size = 0;
+            head = new ListNode(null);
+        }
+
+        public void addAll(List<ItemModel> itemList) {
```
Comment (intelligencegear) [inline demo/movielens/online/src/main/java/com/dmetasoul/metaspore/demo/movielens/diversify/diversifier/impl/MaximalMarginalRelevanceDiversifier.java:176]: We should make sure the `itemList` is an instance of ArrayList.

File: demo/movielens/online/src/main/java/com/dmetasoul/metaspore/demo/movielens/diversify/diversifier/impl/MaximalMarginalRelevanceDiversifier.java [hunk 3]
```diff
@@ -54,8 +58,8 @@ public List<ItemModel> diverse(RecommendContext recommendContext,
         // compute the genre in the window
         HashMap<String, Integer> genreInWindow = new HashMap();
         HashMap<String, Integer> genreSplitedInWindow = new HashMap<>();
-
         // start diverse
+        ListNode itemNode = itemLinkList.head.next;
```
Comment (intelligencegear) [inline demo/movielens/online/src/main/java/com/dmetasoul/metaspore/demo/movielens/diversify/diversifier/impl/MaximalMarginalRelevanceDiversifier.java:62]: We should check the `itemLinkList`, `itemLinkList.head` , `itemLinkList.next` are not `null` here.

File: demo/movielens/online/src/main/java/com/dmetasoul/metaspore/demo/movielens/diversify/diversifier/impl/MaximalMarginalRelevanceDiversifier.java [hunk 4]
```diff
@@ -29,23 +29,27 @@
...
                                    List<ItemModel> itemModels,
                                    Integer window,
                                    Integer tolerance
     ) {
+        if (itemModels.size() <= window) return itemModels;
         Double lambda = recommendContext.getLambda();
         if (lambda == null) {
             lambda = DEFAULT_LAMBDA;
         }
+        LinkList itemLinkList = new LinkList();
```
Comment (intelligencegear) [inline demo/movielens/online/src/main/java/com/dmetasoul/metaspore/demo/movielens/diversify/diversifier/impl/MaximalMarginalRelevanceDiversifier.java:45]: Please change LinkList to LinkedList.

**Review by intelligencegear — CHANGES_REQUESTED** [review]

**Review by intelligencegear — CHANGES_REQUESTED** [review]
Please add unit test code for corner cases.

**Review by intelligencegear — APPROVED** [review]
